### PR TITLE
Hotfix: Enhancements in Testing and MongoDB Connection

### PR DESCRIPTION
**Pull request by PHANTOMS0308**
#### Overview

This pull request focuses on addressing key challenges our team has been facing around our testing environment and the MongoDB connection. The central theme behind these changes is to refine our codebase by addressing identified issues and adopting modern best practices.

###### Problem:
When configuring our Jest test environment, we exclusively set it to `jsdom`. This decision resulted in our backend developers encountering the `TextEncoder is not defined` error, which wouldn’t arise in a standard Node.js testing environment in Jest. However, setting our environment only to Node would disrupt our frontend tests.

###### Solution:
We explored three potential solutions to this dilemma:

1. **Top-Line Comment in Jest**: Declaring the test environment before each test file runs. This method would be tedious and not particularly elegant.
2. **Use of `{ projects: [ ] }` in Jest Config**: A common approach in monorepo. However, this introduced complexities with transforming TypeScript to JavaScript and compatibility issues with `next/jest`. No solutions were readily available from the Next.js community for this.
3. **Naming Convention for Test Files**: The chosen method due to its simplicity and scalability. Test files now follow a `**/*.client.test.tsx` or `**/*.server.test.ts` naming pattern, allowing us to distinguish the testing environments easily.

To execute the tests:

- For server tests: `npm run test:server`
- For client tests: `npm run test:client`

To use flags, such as `--watch`, append it like so: `npm run test:server -- --watch`. You can also add new npm scripts as per your requirements.

###### Problem:
Our initial MongoDB client connection was hastily constructed to kick off our sprint. It stored the cache in `globalThis`—not in line with modern Node.js best practices. The TypeScript errors and lack of error handling posed risks for our serverless functions on Vercel deployment. Also, the absence of a disconnect function made server tests hang, which was counterproductive.

###### Solution:
- We moved away from using `globalThis` to leverage module-level closure.
- Comprehensive error handling was added, especially vital for serverless functions.
- A disconnect function was introduced to ensure tests do not hang.
  
Note: As a consequence of these changes, the MongoDB client file is renamed to `mongoClient.ts`.

##### 3. Environment Variables

Since we're leveraging Next.js, it has its built-in mechanism for managing environment variables, thus making `dotenv` redundant in our setup.

For setting up the environment specific to testing:
- We have provided an `example.env.test.local` file alongside the `example.env.local`.
- To set up your testing environment, copy `example.env.test.local` to a new file named `.env.test.local` and adjust the values as required for testing.
- Similarly, for your local development environment settings, copy `example.env.local` to `.env.local`.

#### Modified Files

- `mongoClient.ts`: Refactored MongoDB connection logic and filename.
- Jest configurations: Split and tailored for client and server-side tests.
- Various test files: Renamed to fit the new naming convention.

#### Package.json

No additional packages were introduced in this hotfix. If any team member is using Babel configurations, it's advised to remove them to avoid conflicts and redundant dependencies. Let Next.js and its built-in SWC handle the transformations.

#### Action Required

1. Please incorporate these changes into your local environment and thoroughly test them to ensure no additional issues arise.
2. Prior to your next pull, ensure your configuration does not introduce new errors, especially if you've made tweaks or additions.
3. Be aware of the renamed MongoDB client file (`mongoClient.ts`) in your integrations.
4. Make sure you have `.env.test.local` setup correctly if your test involves using environment variables.

#### Request for Review

Feedback is invaluable. Please review the updated logic in `mongoClient.ts`, especially the error handling, and the new testing configurations. Your thoughts on the functionality and potential areas of improvement are highly appreciated.

Thank you.

### CODE REVIEW

File: lib/mongoClient.ts
Comment (VijayIyer) [inline lib/mongoClient.ts]: Might want to specify some sort of message here. 
I checked out the pr locally and at first ran `npm run test:server` without creating an `.env.test.local` and jest crashed the test at this line. 
Ideally a user / dev shouldn't make the basic mistake I made above but the Error message might still be worth changing to make it more clear

**Review by VijayIyer — APPROVED** [review]
I added a minor comment on the lib/mongoClient.ts file but that is just a minor suggestion, this PR looks good to be merged

File: lib/mongoClient.ts
Comment (PHANTOMS0308) [reply lib/mongoClient.ts]: Sure, thanks for catching that for me, let me commit one more time. Probably I was just putting here to make Typescript warning disappear, then forget about changing the `...` into a message. 🤣

**Comment by PHANTOMS0308** [comment]
> I added a minor comment on the lib/mongoClient.ts file but that is just a minor suggestion, this PR looks good to be merged

Cool, nice to hear it, I am also waiting for Alex @chenyou-H  to review and test, since I want Lynnda to follow and fix her code based on the latest dev after Alex merge

**Review by chenyou-H — APPROVED** [review]

**Comment by chenyou-H** [comment]
This test config setting  for the client and the server. The test runs well for my API test in my environment. I don't need to manually edit my node module code anymore

**Comment by chenyou-H** [comment]
Good to merge

**Comment by chenyou-H** [comment]
LGTM!!

**Pull request merged**
